pkg/controller/monitor/cnr_indicator.go

```diff
@@ -0,0 +1,212 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package monitor
+
+import (
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+
+	"github.com/kubewharf/katalyst-api/pkg/apis/node/v1alpha1"
+	"github.com/kubewharf/katalyst-core/pkg/config/generic"
+	"github.com/kubewharf/katalyst-core/pkg/metrics"
+	"github.com/kubewharf/katalyst-core/pkg/util/native"
+	"github.com/kubewharf/katalyst-core/pkg/util/qos"
+)
+
+const (
+	metricsNameCNRReportAnomaly  = "cnr_report_anomaly"
+	metricsNameCNRReportLantency = "cnr_report_latency"
+)
+
+const (
+	// reasonNumaExclusiveAnomaly is the reason for numa exclusive anomaly
+	// when numa_binding and numa_exclusive are both set
+	// the pod with numa_binding and numa_exclusive shares the numa with other pods
+	reasonNumaExclusiveAnomaly = "NumaExclusiveAnomaly"
+	// reasonNumaAllocatableSumAnomaly is the reason for numa allocatable sum anomaly
+	// when the node's sum of numa allocatable is not equal to the node allocatable
+	reasonNumaAllocatableSumAnomaly = "AllocatableSumAnomaly"
+	// reasonPodAllocationSumAnomaly is the reason for pod allocation sum anomaly
+	// when the numa's sum of pod allocation is greater than the numa allocatable
+	reasonPodAllocationSumAnomaly = "PodAllocationSumAnomaly"
+)
+
+// checkNumaExclusiveAnomaly checks whether exist the pod with numa_binding and numa_exclusive shares the numa with other pods
+func (ctrl *CNRMonitorController) checkNumaExclusiveAnomaly(cnr *v1alpha1.CustomNodeResource) bool {
+	qosConf := generic.NewQoSConfiguration()
+	for _, socket := range cnr.Status.TopologyZone {
+		for _, numa := range socket.Children {
+			if numa.Type != v1alpha1.TopologyTypeNuma {
+				// only check numa
+				continue
+			}
+			podNum := len(numa.Allocations)
+			for _, allocation := range numa.Allocations {
+				key := allocation.Consumer
+				namespace, podname, _, err := native.ParseUniqObjectUIDKey(key)
+				if err != nil {
+					klog.Errorf("[CNRIndicatorNumaExclusiveAnomaly] failed to parse uniq object uid key %s", key)
+					continue
+				}
+				pod, err := ctrl.podLister.Pods(namespace).Get(podname)
+				if err != nil {
+					klog.Errorf("[CNRIndicatorNumaExclusiveAnomaly] failed to get pod %s", key)
+					continue
+				}
+				if qos.IsPodNumaBinding(qosConf, pod) && qos.IsPodNumaExclusive(qosConf, pod) && podNum > 1 {
+					return true
+				}
+			}
+		}
+	}
+	return false
+}
+
+// checkNumaAllocatableSumAnomaly checks whether the node's sum of numa allocatable is not equal to the node allocatable
+func (ctrl *CNRMonitorController) checkNumaAllocatableSumAnomaly(cnr *v1alpha1.CustomNodeResource) bool {
+	node, err := ctrl.client.KubeClient.CoreV1().Nodes().Get(ctrl.ctx, cnr.Name, metav1.GetOptions{ResourceVersion: "0"})
```

Comment:
Using an informer/lister to obtain nodes is preferable in most situations, as it's faster than clientset due to its local cache. Clientset involves remote calls to the API server, which slows down the process.

Reply:
OK, I will fix this.